src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ClusterMonitorActor.scala

```diff
@@ -0,0 +1,291 @@
+  private[monitor] case class QueryForCluster() extends ClusterMonitorMessage
+  private[monitor] case class ReadyCluster(publicIP: String) extends ClusterMonitorMessage
+  private[monitor] case class NotReadyCluster(status: ClusterStatus) extends ClusterMonitorMessage
+  private[monitor] case class FailedCluster(code: Int, message: Option[String]) extends ClusterMonitorMessage
+  private[monitor] case class DeletedCluster() extends ClusterMonitorMessage
+  private[monitor] case class ShutdownActor(notifyParentMsg: Option[Any] = None) extends ClusterMonitorMessage
+
+  /** Used for tracking internal exceptions in ClusterMonitorActor. This exception is not returned to users. */
+  private[monitor] case class ClusterMonitorException(errMsg: String) extends LeoException(errMsg)
+}
+
+/**
+  * An actor which monitors the status of a Cluster. Periodically queries Google for the cluster status,
+  * and acts appropriately for Running, Deleted, and Failed clusters.
+  * @param cluster the Cluster to monitor
+  * @param gdDAO the Google dataproc DAO
+  * @param dbRef the DB reference
+  */
+class ClusterMonitorActor(val cluster: Cluster,
+                          val monitorConfig: MonitorConfig,
+                          val gdDAO: DataprocDAO,
+                          val dbRef: DbReference) extends Actor with LazyLogging {
+  import context._
+
+  override def preStart(): Unit = {
+    scheduleMonitorPass
+    super.preStart()
+  }
+
+  override def receive: Receive = {
+    case Right(msg) =>
+      logger.debug(s"ClusterMonitorActor: Received message $msg")
+
+      msg match {
+        case ScheduleMonitorPass() => scheduleMonitorPass
+        case QueryForCluster() => getGoogleCluster.value pipeTo self
+        case NotReadyCluster(status) => handleNotReadyCluster(status).value pipeTo self
+        case ReadyCluster(ip) => handleReadyCluster(ip).value pipeTo self
+        case FailedCluster(code, message) => handleFailedCluster(code, message).value pipeTo self
+        case DeletedCluster() => handleDeletedCluster.value pipeTo self
+        case ShutdownActor(notifyParentMsg) =>
+          notifyParentMsg.foreach(msg => parent ! msg)
+          stop(self)
+      }
+
+    case Left(exception: Throwable) =>
+      // An error occurred, let the supervisor handle & log it
+      throw exception
+
+    case Failure(exception) =>
+      // An error occurred, let the supervisor handle & log it
+      throw exception
+  }
+
+  private def scheduleMonitorPass: Unit = {
+    system.scheduler.scheduleOnce(monitorConfig.pollPeriod, self, Right(QueryForCluster()))
+  }
+
+  /**
+    * Handles a dataproc cluster which is not ready yet. We don't take any action, just
+    * schedule another monitor pass.
+    * @param status the Google status
+    * @return error or ScheduleMonitorPass
+    */
+  private def handleNotReadyCluster(status: ClusterStatus): EitherT[Future, Throwable, ScheduleMonitorPass] = {
+    EitherT.pure(ScheduleMonitorPass())
```

Comment:
So this is using `EitherT` for error handling. I think it provides a nice way of doing error handling with minimal boilerplate. Some things to keep in mind:

- `EitherT[Future, A, B]` is a fancy way of writing `Future[Either[A, B]]`
- When you `flatMap` an `EitherT[Future, A, B]`, you operate on the `B`. In other words you're unwrapping 2 layers of effects, instead of 1.
- In this case I'm using `EitherT[Future, Throwable, A]`. Whether I have a `Either[Throwable, A]`, a `Future[A]`, or just an `A`, I can conveniently lift _all_ of those types into an `EitherT[Future, Throwable, A]`.